File: libraries/OclIterator.java
import java.util.*;
import java.util.HashMap;
import java.util.Collection;
import java.util.List;
import java.util.ArrayList;
import java.util.Set;
import java.util.HashSet;
import java.util.TreeSet;
import java.util.Collections;
import java.util.function.Function;


class OclIteratorResult { 
  static OclIteratorResult createOclIteratorResult() { 
    OclIteratorResult result = new OclIteratorResult();
    return result; 
  }

  boolean done = false;
  Object value = null;

  static OclIteratorResult newOclIteratorResult(Object v) 
  { OclIteratorResult res = createOclIteratorResult();
    res.value = v;
    if (v == null) 
    { res.done = true; }
    else
    { res.done = false; }
    return res;
  } 

}


class OclIterator { static ArrayList<OclIterator> OclIterator_allInstances = new ArrayList<OclIterator>();

  OclIterator() { OclIterator_allInstances.add(this); }

  static OclIterator createOclIterator() { OclIterator result = new OclIterator();
    return result; }

   int position = 0;
   int markedPosition = 0;
   Function<Integer,Object> generatorFunction = null; 

  ArrayList<Object> elements = (new ArrayList());
  ArrayList<String> columnNames = (new ArrayList<String>());
 
  String ocliteratorId = ""; /* primary */
  static Map< String,OclIterator> OclIterator_index = new HashMap< String,OclIterator>();


  static OclIterator createByPKOclIterator( String ocliteratorIdx)
  { OclIterator result = OclIterator.OclIterator_index.get(ocliteratorIdx);
    if (result != null) { return result; }
    result = new OclIterator();
    OclIterator.OclIterator_index.put(ocliteratorIdx,result);
    result.ocliteratorId = ocliteratorIdx;
    return result; }

  static void killOclIterator( String ocliteratorIdx)
  { OclIterator rem = OclIterator_index.get(ocliteratorIdx);
    if (rem == null) { return; }
    ArrayList<OclIterator> remd = new ArrayList<OclIterator>();
    remd.add(rem);
    OclIterator_index.remove(ocliteratorIdx);
    OclIterator_allInstances.removeAll(remd);
  }


  public  boolean hasNext()
  {
     boolean result = false;
    if (position >= 0 && position < elements.size())
    {
      result = true;
    }
    else {
      result = false;
    }
    return result;
  }


  public boolean isAfterLast() 
  { boolean result = false; 
    if (position > elements.size())
    { return true; } 
    return result; 
  } 

  public boolean isBeforeFirst() 
  { boolean result = false; 
    if (position <= 0)
    { return true; } 
    return result; 
  } 

  public boolean hasPrevious()
  {
     boolean result = false;
    if (position > 1 && position <= elements.size() + 1)
    {
      result = true;
    }
    else {
      result = false;
    }
    return result;
  }


  public int nextIndex()
  {
    int result = 0;
    result = position + 1;
    return result;
  }


  public int previousIndex()
  {
    int result = 0;
    result = position - 1;
    return result;
  }


  public void moveForward()
  {
    position = position + 1;
  }


  public void moveBackward()
  {
    position = position - 1;
  }


  public void moveTo( int i)
  {
    position = i;
  }

  public void setPosition(int i)
  {
    position = i;
  }

  public void markPosition()
  {
    markedPosition = position;
  }

  public void movePosition(int i)
  {
    position = i + position;
  }

  public void moveToMarkedPosition()
  {
    position = markedPosition;
  }

  public void moveToFirst()
  {
    position = 1;
  }

  public void moveToLast()
  {
    position = elements.size();
  }


  public void moveToStart()
  {
    position = 0;
  }


  public void moveToEnd()
  {
    position = elements.size() + 1;
  }


  public static OclIterator newOclIterator_Sequence(ArrayList sq)
  { OclIterator ot = null;
    ot = OclIterator.createOclIterator();
    ot.elements = sq;
    ot.position = 0;
    return ot; 
  }

  public static OclIterator newOclIterator_String(String str)
  { OclIterator ot = null;
    ot = OclIterator.createOclIterator();
    ot.elements.addAll(Ocl.split(str, "[ \n\t\r]+"));
    ot.position = 0;
    return ot; 
  }

  public static OclIterator newOclIterator_String_String(String str, String seps)
  { OclIterator ot = null;
    ot = OclIterator.createOclIterator();
    ot.elements.addAll(Ocl.split(str, "[" + seps + "]+"));
    ot.position = 0;
    return ot; 
  }


  public static OclIterator newOclIterator_Set(Collection st)
  { ArrayList elems = new ArrayList(); 
    elems.addAll(st); 
    Collections.sort(elems); 
     OclIterator ot = null;
    ot = OclIterator.createOclIterator();
    ot.elements = elems;
    ot.position = 0;
    return ot; 
  }

  public static OclIterator newOclIterator_Function(Function<Integer,Object> f)
  { OclIterator ot = null;
    ot = OclIterator.createOclIterator();
    ot.generatorFunction = f; 
    ot.position = 0;
    return ot; 
  }

  public OclIterator trySplit()  
  { ArrayList firstpart = Ocl.subrange(elements,1,position-1); 
    elements = Ocl.subrange(elements, position); 
    position = 0;
	markedPosition = 0; 
    return OclIterator.newOclIterator_Sequence(firstpart); 
  } 

  public Object getCurrent()
  {
    Object result = null;
    if (position > 0 && position <= elements.size())
    { result = ((Object) elements.get(position - 1)); } 
    return result;
  }


  public void set(Object x)
  {
    elements.set(position - 1,x);
  }


  public void insert(Object x)
  { elements.add(position-1,x); }


  public void remove()
  { elements.remove(position - 1); }

  public boolean tryAdvance(Function<Object,Object> f) 
  { if (position + 1 <= elements.size())  
    { Object x = this.next();   
      f.apply(x);  
      return true; 
    } 
    return false; 
  }  
  
  public void forEachRemaining(Function<Object,Object> f)  
  { ArrayList remainingElements = Ocl.subrange(elements, position); 
    for (Object x : remainingElements)
    { f.apply(x); } 
  }  


  public OclIteratorResult nextResult()
  { if (generatorFunction == null) 
    { Object v = next();
      return OclIteratorResult.newOclIteratorResult(v);
    }
 
    Object res = generatorFunction.apply(new Integer(position)); 
    position++; 
    if (position <= elements.size())  
    { set(res); } 
    else
    { elements.add(res); } 
    return OclIteratorResult.newOclIteratorResult(res); 
  } 
      
  public Object next()
  { if (generatorFunction != null) 
    { Object res = generatorFunction.apply(new Integer(position)); 
      position++; 
      if (position <= elements.size())  
      { set(res); } 
      else
      { elements.add(res); } 
      return res; 
    } 
     
    Object result = null;
    moveForward();
    return getCurrent();
  }


  public Object previous()
  {
    Object result = null;
    moveBackward();
    return getCurrent();
  }

  public Object at(int i)
  { return elements.get(i-1); } 

  public int length()
  { return elements.size(); } 

  public int getPosition()
  { return position; } 
  
  public int getColumnCount()  
  { return columnNames.size(); } 

  public String getColumnName(int i)  
  { if (i > 0 && i <= columnNames.size())
    { return columnNames.get(i-1); } 
    return null; 
  } 

  public Object getCurrentFieldByIndex(int i)  
  { String fld = columnNames.get(i-1); 
    Map curr = (Map) this.getCurrent();
    if (curr != null && fld != null)  
    { return curr.get(fld); } 
    return null; 
  } 

  public void setCurrentFieldByIndex(int i, Object v)  
  { String fld = columnNames.get(i-1); 
    Map curr = (Map) this.getCurrent();
    if (curr != null && fld != null)  
    { curr.put(fld,v); } 
  } 

  public void close()
  { position = 0;
    markedPosition = 0;
    elements = (new ArrayList());
    columnNames = new ArrayList<String>(); 
  } 
  
  /*
  public static void main(String[] args)
  { ArrayList lst = new ArrayList(); 
    lst.add(1); lst.add(2); lst.add(3); lst.add(4); lst.add(5); 
    OclIterator iter1 = OclIterator.newOclIterator_Sequence(lst); 
	System.out.println(iter1.elements); 
	iter1.setPosition(3); 
	OclIterator iter2 = iter1.trySplit(); 
	System.out.println(iter1.elements); 
	System.out.println(iter2.elements); 
	
	Function<Object,Object> ff = (y)->{ System.out.println(y); return y; }; 
	iter1.forEachRemaining(ff); 
  } */ 

}

File: libraries/OclIterator.cpp
template<class T>
bool OclIterator<T>::hasNext()
{
    bool result = false;
    if (position >= 0 && position < elements->size())
    {
        result = true;
    }
    else
    {
        result = false;
    }
    return result;
}

template<class T>
bool OclIterator<T>::hasPrevious()
{
    bool result = false;
    if (position > 1 && position <= elements->size() + 1)
    {
        result = true;
    }
    else
    {
        result = false;
    }
    return result;
}

template<class T>
bool OclIterator<T>::isAfterLast()
{
    bool result = false;
    if (position > elements->size())
    {
        result = true;
    }
    else
    {
        result = false;
    }
    return result;
}

template<class T>
bool OclIterator<T>::isBeforeFirst()
{
    bool result = false;
    if (position < 1)
    {
        result = true;
    }
    else
    {
        result = false;
    }
    return result;
}

template<class T>
int OclIterator<T>::nextIndex()
{
    int result = 0;
    result = position + 1;
    return result;
}

template<class T>
int OclIterator<T>::previousIndex()
{
    int result = 0;
    result = position - 1;
    return result;
}

template<class T>
void OclIterator<T>::moveForward()
{
    int pre_position0 = position;
    //  if (position + 1 > elements->size())) { return; } 
    this->setPosition(pre_position0 + 1);
}

template<class T>
void OclIterator<T>::moveBackward()
{
    int pre_position1 = position;
    //  if (position <= 1)) { return; } 
    this->setPosition(pre_position1 - 1);
}

template<class T>
void OclIterator<T>::moveTo(int i)
{ //  if ((!(0 <= i) || i > elements->size() + 1))) { return; } 
    this->setPosition(i);
}

template<class T>
void OclIterator<T>::movePosition(int i)
{
    int pos = this->getPosition();
    this->setPosition(i + pos);
}

template<class T>
void OclIterator<T>::moveToStart()
{
    this->setPosition(0);
}

template<class T>
void OclIterator<T>::moveToEnd()
{
    this->setPosition(this->elements->size() + 1);
}

template<class T>
void OclIterator<T>::moveToFirst()
{
    this->setPosition(1);
}

template<class T>
void OclIterator<T>::moveToLast()
{
    this->setPosition(this->elements->size());
}

template<class T>
OclIterator<T>* OclIterator<T>::newOclIterator_Sequence(vector<T>* sq)
{
    OclIterator<T>* result = NULL;
    OclIterator<T>* ot = new OclIterator<T>();
    ot->setelements(sq);
    ot->setPosition(0);
    ot->markedPosition = 0;
    ot->columnNames = new vector<string>();
    result = ot;
    return result;
}
 
template<class T>
OclIterator<T>* OclIterator<T>::newOclIterator_Set(std::set<T>* st)
{
    OclIterator<T>* ot = new OclIterator<T>();
    vector<T>* elems = UmlRsdsLib<T>::concatenate((new vector<T>()), st);

    ot->elements = UmlRsdsLib<T>::sort(elems);
    ot->setPosition(0);
    ot->markedPosition = 0;
    ot->columnNames = new vector<string>();
    return ot;
} // sort the elements, to account for sorted sets

template<class T>
OclIterator<T>* OclIterator<T>::newOclIterator_String(string st)
{
    OclIterator<string>* res = new OclIterator<string>();
    res->elements = UmlRsdsLib<string>::split(st, "[ \n\t\r]");
    res->position = 0;
    res->markedPosition = 0;
    res->columnNames = new vector<string>();
    return res;
}

template<class T>
OclIterator<T>* OclIterator<T>::newOclIterator_String_String(string st, string sep)
{
    OclIterator<string>* res = new OclIterator<string>();
    res->elements =
        UmlRsdsLib<string>::split(st,
            string("[").append(sep).append(string("]")));
    res->position = 0;
    res->markedPosition = 0;
    res->columnNames = new vector<string>();
    return res;
}
  

template<class T>
OclIterator<T>* newOclIterator_Function(function<T (int)> f)
{
    OclIterator<T>* res = new OclIterator<T>();
    res->elements = new vector<T>(); 
    res->generatorFunction = f;
    res->position = 0;
    res->markedPosition = 0;
    res->columnNames = new vector<string>();

    return res;
}

template<class T>
OclIterator<T> OclIterator<T>::trySplit()
{
  vector<T>* firstpart = UmlRsdsLib<T>::subrange(elements, 1, position - 1);
  elements = UmlRsdsLib<T>::subrange(elements, position, elements->size());
  position = 0;
  markedPosition = 0;
  return OclIterator.newOclIterator_Sequence(firstpart);
}


template<class T>
T OclIterator<T>::getCurrent()
{
    T result;
    if (position < 1 || position > elements->size())
    {
        return result;
    }
    result = ((T)elements->at(position - 1));
    return result;
}

template<class T>
void OclIterator<T>::set(T x)
{
    this->setelements(this->getPosition() - 1, x);
}

template<class T>
void OclIterator<T>::insert(T x)
{
    elements->insert(elements->begin() + (position - 1), x);
}

template<class T>
void OclIterator<T>::remove()
{
    elements->erase(elements->begin() + (position - 1));
}

template<class T>
OclIteratorResult<T> OclIterator<T>::nextResult()
{
    if (generatorFunction == NULL)
    {
        OclIterator<T>* ocliteratorx = this;
        ocliteratorx->moveForward();
        return OclIteratorResult<T>::newOclIteratorResult(ocliteratorx->getCurrent());
    } 
    T r = this.generatorFunction(this->position); 
    this->position = this->position + 1;
    if (this->position <= this->elements->size())
    {
       this->set(r);
    } 
    else
    {
       this->elements->push_back(r);
    } 
    return OclIteratorResult<T>::newOclIteratorResult(r);
}

template<class T>
T OclIterator<T>::next()
{
    OclIterator<T>* ocliteratorx = this;
    ocliteratorx->moveForward();
    return ocliteratorx->getCurrent();
}

template<class T>
bool OclIterator<T>::tryAdvance(function<void*(void*)> f)
{
    if (this->position + 1 <= this->elements->size())
    {
        void* x = this->next();
        f(x);
        return true;
    }
    return false;
}


template<class T>
void OclIterator<T>::forEachRemaining(function<void*(void*)> f)
{
  vector<T>* remainingElements = UmlRsdsLib<T>::subrange(elements, position, elements->size());
  for (int i = 0; i < remainingElements->size(); i++)
  {
      T x = remainingElements->at(i); 
      f(x);
  }
}

template<class T>
T OclIterator<T>::previous()
{
    OclIterator<T>* ocliteratorx = this;
    ocliteratorx->moveBackward();
    return ocliteratorx->getCurrent();
}

template<class T>
T OclIterator<T>::at(int i)
{
    return elements->at(i - 1);
}

template<class T>
int OclIterator<T>::length()
{
    return elements->size();
}

template<class T>
void OclIterator<T>::close()
{
    position = 0;
    markedPosition = 0;
    columnNames = new vector<string>(); 
    elements = new vector<T>();
}

template<class T>
int OclIterator<T>::getColumnCount()
{
    return columnNames->size();
}

template<class T>
string OclIterator<T>::getColumnName(int i)
{
    if (columnNames->size() >= i && i >= 1)
    {
        return columnNames->at(i - 1);
    }
    return "";
}

template<class T>
void* OclIterator<T>::getCurrentFieldByIndex(int i)
{
    string nme = getColumnName(i); 
    T curr = getCurrent(); 
    map<string, void*> m = (map<string, void*>) curr; 
    return m->at(nme);
}

template<class T>
void OclIterator<T>::setCurrentFieldByIndex(int i, void* v)
{
    string nme = getColumnName(i);
    T curr = getCurrent();
    map<string, void*> m = (map<string, void*>) curr;
    m[nme] = v;
}



